docs/proposals/set-ref.md

```diff
@@ -0,0 +1,144 @@
+<!-- BEGIN MUNGE: UNVERSIONED_WARNING -->
+
+<!-- BEGIN STRIP_FOR_RELEASE -->
+
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# ControllerRef proposal
+
+Author: gmarek@
+Last edit: 2016-05-06
+Status: raw
+
+Approvers:
+- [ ] briangrant
+- [ ] dbsmith
+
+**Table of Contents**
+
+- [Goal of SetReference](#goal-of-setreference)
+- [Non goals](#non-goals)
+- [API and semantic changes](#api-and-semantic-changes)
+- [Upgrade/downgrade procedure](#upgradedowngrade-procedure)
+- [Orphaning/adoption](#orphaningadoption)
+- [Implementation plan (sketch)](#implementation-plan-sketch)
+- [Considered alternatives](#considered-alternatives)
+
+# Goal of SetReference
+
+Main goal of *setReference* effort is to solve a problem of overlapping controllers that fight over some resources (e.g. *ReplicaSets* fighting with *ReplicationControllers* over *Pods*), which cause serious [problems](https://github.com/kubernetes/kubernetes/issues/24433) such as exploding memory of Controller Manager.
+
+We don’t want to have (just) an in-memory solution, as we don’t want a Controller Manager crash to cause massive changes in object ownership in the system. I.e. we need to persist the information about "owning controller".
+
+Secondary goal of this effort is to improve performance of various controllers and schedulers, by removing the need for expensive lookup for all matching "controllers".
+
+# Non goals
+
+Cascading deletion is not a goal of this effort. Cascading deletion will use *ownerReferences*, which is a [separate effort](garbage-collection.md).
+
+There will be some interactions between set and owner references, they will have a similar orphaning/adoption behavior and *setReference* may be treated as one of *ownerReferencess*.
+
+# API and semantic changes
+
+There will be a new API field in the *ObjectMeta* in which we will store a *setReference*:
+
+```
+ObjectMeta {
+    …
+    setReference *SetReference
```

Comment:
@gmarek's original idea was collection. I suggested set, on the argument that it is no more vague than collection (except to former Borg users) and it contains the suffix we're kinda trying to use for API objects (I say kinda because we did ReplicaSet and DaemonSet, but then fell off the wagon with Job and Deployment). You make a good point that "set" might be interpreted as a verb. I'm fine collection. I don't think the other terms you listed are good, because I think we want to covey that this is the set, not the logic that manages the set (which AIUI is part of the reason we moved away from calling collections *Controller, to calling them *Set).